File: Problems/Mathematics/Day-08/soln/prashant_dwivedi/Solution1.java
import java.util.*;
import java.io.*;

public class Solution1 {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {

            String[] first = br.readLine().split(" ");
            int n = Integer.parseInt(first[0]);
            int k = Integer.parseInt(first[1]);

            long[] a = new long[n];
            String[] arr = br.readLine().split(" ");
            for (int i = 0; i < n; i++) {
                a[i] = Long.parseLong(arr[i]);
            }

            Arrays.sort(a);

            int m = n - k;
            int mid = (m - 1) / 2;

            long left = a[mid];
            long right = a[n - 1 - mid];

            System.out.println(right - left + 1);
        }
    }
}

/*
-------------------- SIMPLE DOCUMENTATION --------------------

What this code does:
- Reads positions of bars on a street
- Allows closing at most k bars
- Finds how many house positions can give minimum total distance

Core idea:
- To minimize distance, always remove the farthest bars
- After removing k bars, only (n - k) bars matter
- The best position is always at the median

Why sorting:
- Makes it easy to remove farthest bars
- Helps find the median range directly

How answer is computed:
- Find the valid median range [left, right]
- Every integer in this range gives the same minimum distance
- Count = right - left + 1

Time Complexity:
- O(n log n) due to sorting

Submission:
https://codeforces.com/contest/2098/submission/356145066
----------------------------------------------------------
*/

File: Problems/Binary-Search/Day-07/sol/Ishan_Raj_Singh/Solution1.py
"""
Approach: Two Pointers approach
Submission Link: https://codeforces.com/problemset/submission/1538/356144559
"""

def solve_two_pointers():
    n, l, r = map(int, input().split())
    a = list(map(int, input().split()))
    
    a.sort()
    count = 0
    
    for i in range(n):
        min_sum = l - a[i]
        max_sum = r - a[i]
        
        left = i + 1
        right = n - 1
        
        lo, hi = i + 1, n
        while lo < hi:
            mid = (lo + hi) // 2
            if a[mid] < min_sum:
                lo = mid + 1
            else:
                hi = mid
        left_bound = lo
        
        lo, hi = i + 1, n
        while lo < hi:
            mid = (lo + hi) // 2
            if a[mid] <= max_sum:
                lo = mid + 1
            else:
                hi = mid
        right_bound = lo - 1
        
        if left_bound < n and right_bound >= i + 1 and left_bound <= right_bound:
            count += (right_bound - left_bound + 1)
    
    print(count)

t = int(input())
for _ in range(t):
    solve_two_pointers()
